sympy/stats/stochastic_process_types.py

```diff
@@ -0,0 +1,315 @@
+        if not isinstance(process, DiscreteMarkovChain):
+            raise ValueError("Currently only DiscreteMarkovChain "
+                                "support StateSpace.")
+        state_space = _set_converter(state_space)
+        return Basic.__new__(cls, process, state_space)
+
+    process = property(lambda self: self.args[0])
+    state_space = property(lambda self: self.args[1])
+
+class DiscreteMarkovChain(DiscreteTimeStochasticProcess):
+    """
+    Represents discrete Markov chain.
+
+    Parameters
+    ==========
+
+    sym: Symbol
+    state_space: Set
+        Optional, by default, S.Reals
+    trans_probs: Matrix/ImmutableMatrix/MatrixSymbol
+        Optional, by default, None
+
+    Examples
+    ========
+
+    >>> from sympy.stats import DiscreteMarkovChain, TransitionMatrix
+    >>> from sympy import Matrix, MatrixSymbol, Eq
+    >>> from sympy.stats import P
+    >>> T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])
+    >>> Y = DiscreteMarkovChain("Y", [0, 1, 2], T)
+    >>> YS = DiscreteMarkovChain("Y")
+    >>> Y.state_space
+    {0, 1, 2}
+    >>> Y.trans_probs
+    Matrix([
+    [0.5, 0.2, 0.3],
+    [0.2, 0.5, 0.3],
+    [0.2, 0.3, 0.5]])
+    >>> TS = MatrixSymbol('T', 3, 3)
+    >>> P(Eq(YS[3], 2), Eq(YS[1], 1) & TransitionMatrix(YS, TS))
+    T[0, 2]*T[1, 0] + T[1, 1]*T[1, 2] + T[1, 2]*T[2, 2]
+    >>> P(Eq(Y[3], 2), Eq(Y[1], 1)).round(2)
+    0.36
+    """
+
+    index_set = S.Naturals0
+
+    def __new__(cls, sym, state_space=S.Reals, trans_probs=None):
+        sym = _symbol_converter(sym)
+        state_space = _set_converter(state_space)
+        if trans_probs != None:
+            trans_probs = _matrix_checks(trans_probs)
+        return Basic.__new__(cls, sym, state_space, trans_probs)
+
+    @property
+    def trans_probs(self):
+        """
+        Transition probabilities of discrete Markov chain,
+        either an instance of Matrix or MatrixSymbol.
+        """
+        return self.args[2]
+
+    def probability(self, condition, given_condition, **kwargs):
+        """
+        Handles probability queries for discrete Markov chains.
+
+        Parameters
+        ==========
+
+        condition: Relational
+        given_condition: Relational/And
+
+        Returns
+        =======
+
+        Probability
+            If the transition probabilities are not available
+        Expr
+            If the transition probabilities is MatrixSymbol or Matrix
+
+        Note
+        ====
+
+        Any information passed at the time of query overrides
+        any information passed at the time of object creation like
+        transition probabilities, state space.
+
+        Pass the transition matrix using TransitionMatrix and state space
+        using StateSpace in given_condition using & or And.
+        """
+
+        # extracting transition matrix and state space
+        trans_probs, state_space = self.trans_probs, self.state_space
+        if isinstance(given_condition, And):
```

Comment:
definitely a bad choice, you are limiting the expressions to And only. You should try to use "ask" in SymPy, which supports logical inference.

I'm not sure if this is simple. Can you have a look?

Reply:
I think that even if we use `ask` it will return either `True` or `False`. So to pass transition matrix or state space via `given_condition`, we will have to use `And`. Infact, if try to make out a meaning of something like, `Or(Eq(Probability(Eq(Y[0], 0)), S(1)/4), TransitionMatrix(Y, TO))`, then, we will lead to a conflict that whether to use, `Eq(Probability(Eq(Y[0], 0)), S(1)/4)` or `TransitionMatrix(Y, TO)`. `And` implies that we should use all the `given_condition`. 
For making `ask` work, I will need to make a class that will inherit `AssumptionKeys` and then define an extra keys in it like, `transition_matrix` and `state_space`. That also sounds good. However, which binary operator we should use if not `And`. I think using any other will just lead to conflicts. Isn't it?
Please let me know if you want something else.

Reply:
Or do you want something like this,
```
>>> X = DiscreteMarkovChain('X')
>>> TransitionMarix(X, matrix)
>>> ask(Q.transition_matrix(X, matrix))
True
```
But by the above way, transition probabilities aren't passed via `given_condition`, they have to be assumed separately, which conflicts with the API design. 
Please give me an example to clearly understand your view.

Reply:
Hi, maybe it's too complicated with "ask". Let's keep it the way you have written the code right now.

My concern is that there are expressions that are not really "And" objects, which may be more complicated. But let's ignore them for now.

Reply:
The purpose of ask is to try to get a boolean value from the relations. This code is attempting to extract the transition probabilities from the relations so they need to be relations that don't evaluate to a boolean.

The code here will always look weird if if it needs to support this API where the transition probabilities can be provided as part of the query. There should be an independent API for that use case if it is useful.